bob-access/src/credentials.rs

```diff
@@ -59,66 +74,199 @@ impl CredentialsKind {
+    pub fn build(self) -> RequestCredentials {
+        RequestCredentials {
             address: self.address,
-            kind: self.kind.clone(),
+            kind: self.kind,
         }
     }
 }
 
+const RESOLVE_THRESHOLD: u32 = 3;
+
+#[derive(Debug, Clone)]
+enum ResolveState {
+    Resolved(u32),
+    InProgress,
+}
+
+impl ResolveState {
+    fn update(&mut self, unresolved: bool) -> bool {
+        match self {
+            ResolveState::Resolved(n) => {
+                if unresolved {
+                    *n = n.wrapping_add(1);
+                    *n >= RESOLVE_THRESHOLD
+                } else {
+                    *n = 0;
+                    false
+                }                
+            },
+            ResolveState::InProgress => false,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DeclaredCredentials {
+    address: Vec<SocketAddr>,
+    hostname: Option<String>,
+    kind: CredentialsKind,
+}
+
+impl DeclaredCredentials {
+    pub fn internode_builder(node_name: impl Into<String>) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::InterNode(node_name.into()))
+    }
+
+    pub fn token_builder(token: impl Into<String>) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::Token(token.into()))
+    }
+
+    pub fn userpass_builder(username: impl Into<String>, password: impl Into<String>,) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::Basic {
+            username: username.into(),
+            password: password.into(),
+        })
+    }
+
+    pub fn ip(&self) -> &Vec<SocketAddr> {
+        &self.address
+    }
+
+    pub fn replace_addresses(&mut self, addresses: Vec<SocketAddr>) {
+        self.address = addresses;
+    }
+
+    pub fn hostname(&self) -> &Option<String> {
+        &self.hostname
+    }
+
+    pub fn validate_internode(&self) -> bool {
+        (!self.address.is_empty() || self.hostname.is_some()) &&
+        self.kind.is_internode()
+    }
+
+    pub fn kind(&self) -> &CredentialsKind {
+        &self.kind
+    }
+}
+
+#[derive(Debug)]
+pub struct DeclaredCredentialsBuilder {
+    kind: CredentialsKind,
+    address: Vec<SocketAddr>,
+    hostname: Option<String>,
+}
+
+impl DeclaredCredentialsBuilder {
+    fn new(kind: CredentialsKind) -> Self {
+        Self {
+            kind,
+            address: Vec::new(),
+            hostname: None,
+        }
+    }
+
+    pub fn with_address(mut self, address: SocketAddr) -> Self {
+        self.address = vec![address];
+        self
+    }
+
+    pub fn with_hostname(mut self, hostname: String) -> Self {
+        self.hostname = Some(hostname);
+        self
+    }
+
+    pub fn build(self) -> DeclaredCredentials {
+        DeclaredCredentials {
+            address: self.address,
+            kind: self.kind,
+            hostname: self.hostname,
+        }
+    }
+}
+
+const RESOLVE_SLEEP_DUR_MS: u64 = 1000;
```

Comment:
This should not be a constant. I suggest to use `hostname_resolve_dur_ms` also for protection period.